Add Documents list element

amitaibu opened the pull request:
- Add Documents list paragraph
- `wrapContainerWide`, now accepts an optional background color. If no color then a transparent background is added and no `py` is added, so `verticalSpacing` behaves normally.
- Added `buildElementItemsWithViewMore` as a generic way of adding items that can be hidden with "View more" button

| Style guide | Node view | Node view after "View more" | Node edit |
| -- | -- | -- | -- |
| ![image](https://user-images.githubusercontent.com/125707/223236166-e3f4277c-d56b-471c-9f83-87802399109e.png) | ![image](https://user-images.githubusercontent.com/125707/223239288-bc4d93a7-8f03-4486-826a-08d45bc0e77c.png) | ![image](https://user-images.githubusercontent.com/125707/223239530-f6f45c36-c85c-46dc-a052-0dbefa74f9f6.png) |![image](https://user-images.githubusercontent.com/125707/223239852-9cc71f83-6d28-48b6-a379-6f9cb6c9c6da.png) |

amitaibu commented:
- Added tests.
- Added to migration.

![image](https://user-images.githubusercontent.com/125707/223533786-a5fbb0f2-0936-4487-8452-8597539f8ed5.png)

mariano-dagostino reviewed (changes requested):
@amitaibu Looks good. I finished my review with a few improvements.

amitaibu commented on web/modules/custom/server_general/src/ElementTrait.php:
```diff
@@ -100,6 +100,37 @@ protected function buildElementHeroImage(array $image, string $title, string $su
+    $elements[] = $this->wrapTextColor($subtitle, 'dark-gray');
+
+    // Items and "View more" button.
+    $button = $this->buildButton($this->t('View more'), Url::fromUserInput('#'));
+    $elements[] = $this->buildElementItemsWithViewMore($documents, $button, 2);
```
I could also remove the view more text, and always use that. My concern was if we'd like to show different kind of button (e.g. primary Vs secondary). But let's be lazy - I'll "hardcode" the button.

Extending projects can change it if needed.

amitaibu commented on web/modules/custom/server_general/src/Plugin/EntityViewBuilder/MediaDocument.php:
```diff
@@ -0,0 +1,47 @@
+   */
+  public function buildCard(array $build, MediaInterface $entity): array {
+    /** @var \Drupal\file\FileInterface $file */
+    $file = $this->getReferencedEntityFromField($entity, 'field_media_file');
```
I don't think it's needed, PEVB is just reaching to the Media's field.

amitaibu commented on web/modules/custom/server_style_guide/src/Controller/StyleGuideController.php:
```diff
@@ -507,6 +510,33 @@ protected function getHeroImage(): array {
+      $items[] = $this->buildCardMediaDocument(
+        $this->getRandomTitle(),
+        // Data URI of a png file.
+        'data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAGsAAAB7CAMAAABEt46lAAAACXBIWXMAAA7zAAAO8wEcU5k6AAAAEXRFWHRUaXRsZQBQREYgQ3JlYXRvckFevCgAAAATdEVYdEF1dGhvcgBQREYgVG9vbHMgQUcbz3cwAAAALXpUWHREZXNjcmlwdGlvbgAACJnLKCkpsNLXLy8v1ytISdMtyc/PKdZLzs8FAG6fCPGXryy4AAACglBMVEUnJycoKCgpKSkqKiorKyssLCwtLS0uLi4vLy8wMDAxMTEyMjIzMzM0NDQ1NTU2NjY3Nzc4ODg5OTk6Ojo7Ozs8PDw9PT0+Pj4/Pz9AQEBBQUFDQ0NERERFRUVGRkZHR0dI ...
```
Ok, if that's the reaction - I could load an existing Media image :angel:

amitaibu commented:
Addressed comments, thanks.

| Node View | Style guide |
| -- | -- |
| ![image](https://user-images.githubusercontent.com/125707/223831089-fcbe0223-207f-4ad1-a085-481900881cc2.png) | ![image](https://user-images.githubusercontent.com/125707/223831196-d8d8dff8-b598-45ed-be07-0ca514e732f4.png)|

amitaibu commented on web/modules/custom/server_style_guide/src/Controller/StyleGuideController.php:
```diff
@@ -507,6 +510,51 @@ protected function getHeroImage(): array {
+   *   Render array.
+   */
+  protected function getDocuments(): array {
+    $media_storage = $this->entityTypeManager()->getStorage('media');
+    $result = $media_storage
+      ->getQuery()
+      ->condition('bundle', 'image')
+      ->range(0, 1)
+      ->execute();
+
+    if (empty($result)) {
+      return ['#markup' => 'No Media image found, please add one to see this element'];
+    }
+
+    $media_id = reset($result);
+    /** @var \Drupal\media\MediaInterface $media */
+    $media = $media_storage->load($media_id);
+    $fid = $media->getSource()->getSourceFieldValue($media);
+
+    $file_storage = $this->entityTypeManager()->getStorage('file');
+    /** @var \Drupal\file\FileInterface $file */
+    $file = $file_storage->load($fid);
```
@mariano-dagostino I think maybe to revert to that scary data URI -- so style guide won't depend on actual content. (Think of this as a `View.elm` -- it shouldn't perform any DB query)

Agree?

amitaibu replied:
https://github.com/Gizra/drupal-starter/pull/445/commits/cc84d21cfefbbaf0a95bbc30c9c6de7b25bae6df

mariano-dagostino commented on web/modules/custom/server_style_guide/src/Controller/StyleGuideController.php:
```diff
@@ -507,6 +510,33 @@ protected function getHeroImage(): array {
+      // Add documents.
+      $items[] = $this->buildCardMediaDocument(
+        $this->getRandomTitle(),
+        // Data URI of a png file. We don't load an existing Media ot keep
+        // the style guide working regardless of site content.
+        'data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAGsAAAB7CAMAAABEt46lAAAACXBIWXMAAA7zAAAO8wEcU5k6AAAAEXRFWHRUaXRsZQBQREYgQ3JlYXRvckFevCgAAAATdEVYdEF1dGhvcgBQREYgVG9vbHMgQUcbz3cwAAAALXpUWHREZXNjcmlwdGlvbgAACJnLKCkpsNLXLy8v1ytISdMtyc/PKdZLzs8FAG6fCPGXryy4AAACglBMVEUnJycoKCgpKSkqKiorKyssLCwtLS0uLi4vLy8wMDAxMTEyMjIzMzM0NDQ1NTU2NjY3Nzc4ODg5OTk6Ojo7Ozs8PDw9PT0+Pj4/Pz9AQEBBQUFDQ0NERERFRUVGRkZHR0dI ...
```
```suggestion
        Url::fromUserInput('/modules/custom/server_migrate/files/drupal-starter.pdf')->toString(),
```

mariano-dagostino replied:
@amitaibu :point_up:

amitaibu replied:
Much nicer!
